src/java.desktop/macosx/native/libosxapp/JNIUtilities.m

```diff
@@ -0,0 +1,95 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "JNIUtilities.h"
+
+NSString* JavaStringToNSString(JNIEnv *env, jstring jstr) {
+    if (jstr == NULL) {
+        return NULL;
+    }
+    jsize len = (*env)->GetStringLength(env, jstr);
+    const jchar *chars = (*env)->GetStringChars(env, jstr, NULL);
+    if (chars == NULL) {
+        return NULL;
+    }
+    NSString *result = [NSString stringWithCharacters:(UniChar *)chars length:len];
+    (*env)->ReleaseStringChars(env, jstr, chars);
+    return result;
+}
+
+jstring NSStringToJavaString(JNIEnv* env, NSString *str) {
+
+    if (str == NULL) {
+       return NULL;
+    }
+    jstring jStr = (*env)->NewStringUTF(env, [str UTF8String]);
+    CHECK_EXCEPTION();
+    return jStr;
+}
+
+/*
+ * These next conversion functions are for file system paths.
+ * The NSString needs to be in de-composed UTF-16 format for the Apple file system
+ * The Java String needs to be in pre-composed UTF-16 format for display by Java.
+ * https://developer.apple.com/library/archive/qa/qa1235/_index.html
+ * has some information on this.
+ */
+
+/*
+ * https://developer.apple.com/documentation/foundation/nsstring/1414559-filesystemrepresentation?language=objc
+ * describes how to get a file system representation as a char* from an NSString
+ * and then using FileManager (!) convert it to an NSString.
+ * But we want an NSString.
+ * So the steps are
+ * 1) Convert to NSString
+ * 2) call [NSString fileSystemRepresentation] which gives us a char*
+ * 3) Convert the returned char* to an NSString using FileManager (is there a better way?)
+ */
+NSString* NormalizedPathNSStringFromJavaString(JNIEnv *env, jstring pathStr) {
+    if (pathStr == NULL) {
+        return nil;
+    }
+    NSString *nsStr = JavaStringToNSString(env, pathStr);
+    if (nsStr == NULL) {
+        return nil;
+    }
+    const char* chs = [nsStr fileSystemRepresentation];
+    int len = strlen(chs);
+    NSString* result = [[NSFileManager defaultManager]
+                  stringWithFileSystemRepresentation:chs length:len];
+    return result;
+}
```

Comment:
Why are we doing:

`java_string -> chars -> NSString -> chars -> [NSFileManager stringWithFileSystemRepresentation]`

?

Why not just get the raw characters form JNI and feed them directly into `[NSFileManager  stringWithFileSystemRepresentation]`, ie:

`java_string -> chars -> [NSFileManager stringWithFileSystemRepresentation]`

and skip the `JavaStringToNSString` step altogether?

Reply:
I tried to explain the odd-ness of this in the comments preceding the function definition.
Near as I could figure out that intermediate call is needed to do the decomposition since the NSFileManager won't do that.
But this is not exactly my area of expertise and there may be a better way. Who is an expert on the intricacies of the macOS file system naming ?

Reply:
How about a comment like:

`/*
 Returns an NSString in decomposed UTF16 format that is compatable with HFS's
 expectation of the UTF16 format for file system paths.
 
 Example string: "/Users/Amélie/"
 
 Java's UTF16 string is                 "/ U s e r s / A m \351 l i e /"
 macOS UTF16 string suitable for HFS is "/ U s e r s / A m e \314 \201 l i e /"
 
 There is no direct API that takes in NSString UTF16 encoded by Java
 and produces NSString UTF16 for HFS, so we first need to decompose it
 into chars (suitable for low level C file APIs), and only then
 create NSString representation of this decomposition back into UTF16 string.
 */`

?

I borrowed the API description from Apple's original `JNFNormalizedNSStringForPath` API, but added an example and different description.